core/trino-main/src/main/java/io/trino/operator/join/unspilled/PageJoiner.java

```diff
@@ -157,19 +157,36 @@ private void processUniqueMappingProbe(LookupSource lookupSource)
         int matches = 0;
         int mismatches = 0;
 
-        do {
-            joinPosition = probe.getCurrentJoinPosition();
-            boolean match = joinPosition >= 0 && lookupSource.isJoinPositionEligible(joinPosition, probe.getPosition(), probe.getPage());
-            matches += match ? 1 : 0;
-            mismatches += match ? 0 : 1;
-            if (match) {
-                pageBuilder.appendRow(probe, lookupSource, joinPosition);
+        if (lookupSource.isJoinPositionAlwaysEligible()) {
+            do {
+                joinPosition = probe.getCurrentJoinPosition();
+                boolean match = joinPosition >= 0;
+                matches += match ? 1 : 0;
+                mismatches += match ? 0 : 1;
+                if (match) {
+                    pageBuilder.appendRow(probe, lookupSource, joinPosition);
+                }
+                else if (probeOnOuterSide) {
+                    pageBuilder.appendNullForBuild(probe);
+                }
             }
-            else if (probeOnOuterSide) {
-                pageBuilder.appendNullForBuild(probe);
+            while (probe.advanceNextPosition() && !pageBuilder.isFull() && !yieldSignal.isSet());
+        }
+        else {
+            do {
```

Comment:
the loop here is almost exactly the same as the above. Is this done to avoid inlining issues? If so, consider adding a comment to avoid someone refactoring this in the future.